### ISSUE TITLE
pytorch inference lead to memory leak in cpu

### ISSUE THREAD

**Issue by 836304831** [pytorch/pytorch#55607]
## 🐛 Bug

I inference using pytorch model, I got memory leak problem, my code as follow:

`
import torch
import torch.nn as nn
from memory_profiler import profile
from memory_profiler import memory_usage

@profile(func=None, stream=open('./resnet.log', 'w'))
def run(net, x):
    net(x)

class ConvVerify(nn.Module):

    def init(self):
        for name, m in self.named_modules():
            # print(name, m)
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.running_mean, 0)
                nn.init.constant_(m.running_var, 1)
                if m.weight is not None:
                    nn.init.constant_(m.weight, 1)
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)

    def __init__(self):
        super(ConvVerify, self).__init__()

        stride = 1
        self.model = nn.Sequential(
            nn.Conv2d(288, 64, (2, 2), stride),
            nn.BatchNorm2d(64),
            nn.ReLU(),
        )
        self.init()

    def forward(self, x):
        return self.model(x)

def begin():
    conv_verify_inst = ConvVerify()
    x = torch.autograd.Variable(torch.rand(1, 288, 7, 7))
    for i in range(3000):
        run(conv_verify_inst, x)

if __name__ == "__main__":
    begin()
`

When run some iterator, the memory will increate 0.1M or 0.2M. For exmple, base memory is 71.8 MiB, finally, the memory increated to 72 MiB

## Environment

 - PyTorch Version (e.g., 1.0): 1.4.0 or 1.8.0 
 - OS (e.g., Linux):  windows10 and mac10.13
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6.2 on windows10 and 3.6.13
 - CUDA/cuDNN version: just cpu
 - GPU models and configuration: just cpu

**Comment by peterjc123** [comment]
I modified your script a little bit.
```py
import gc

import torch
import torch.nn as nn

# You may need to do `pip install mem_top`
from mem_top import mem_top

def run(net, x):
    net(x)

class ConvVerify(nn.Module):
    def init(self):
        for _, m in self.named_modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.running_mean, 0)
                nn.init.constant_(m.running_var, 1)
                if m.weight is not None:
                    nn.init.constant_(m.weight, 1)
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)

    def __init__(self):
        super(ConvVerify, self).__init__()

        stride = 1
        self.model = nn.Sequential(
            nn.Conv2d(288, 64, (2, 2), stride),
            nn.BatchNorm2d(64),
            nn.ReLU(),
        )
        self.init()

    def forward(self, x):
        return self.model(x)

def begin():
    conv_verify_inst = ConvVerify()
    x = torch.autograd.Variable(torch.rand(1, 288, 7, 7))
    # Skip the first iteration as there may be some lazy initializations
    run(conv_verify_inst, x)
    print(mem_top())
    print('-' * 60)
    for i in range(3000):
        run(conv_verify_inst, x)
    del conv_verify_inst
    del x
    gc.collect()
    print(mem_top())
    

if __name__ == "__main__":
    begin()
```

Running the script yields the following output.
```
refs:
2338    <class 'dict'> {140726060042256: <weakref at 0x000002C60B8E2DB0; to 'type' at 0x00007FFD56D1C810 (type)>, 140726060
1075    <class 'list'> [(<function _ntuple.<locals>.parse at 0x000002C6129438B0>, 'aten::_pair'), (<function _ntuple.<local
1053    <class 'dict'> (broken __repr__, type=<class 'dict'>)
885     <class 'list'> ['typename', 'is_tensor', 'is_storage', 'set_default_tensor_type', 'set_rng_state', 'get_rng_state',
731     <class 'dict'> {'arange': <staticmethod object at 0x000002C60D8ABBE0>, 'as_tensor': <staticmethod object at 0x00000
677     <class 'list'> ['\n', 'r"""\n', 'The torch package contains data structures for multi-dimensional\n', 'tensors and
626     <class 'dict'> {'sys': <module 'sys' (built-in)>, 'builtins': <module 'builtins' (built-in)>, '_frozen_importlib':
626     <class 'dict'> {'__new__': <built-in method __new__ of torch.tensortype object at 0x000002C610A63E40>, '__doc__': N
626     <class 'dict'> {'__new__': <built-in method __new__ of torch.tensortype object at 0x000002C610A64040>, '__doc__': N
626     <class 'dict'> {'__new__': <built-in method __new__ of torch.tensortype object at 0x000002C610A64240>, '__doc__': N

bytes:
36960    {140726060042256: <weakref at 0x000002C60B8E2DB0; to 'type' at 0x00007FFD56D1C810 (type)>, 140726060
36960    {'sys': <module 'sys' (built-in)>, 'builtins': <module 'builtins' (built-in)>, '_frozen_importlib':
36960    (broken __repr__, type=<class 'dict'>)
36960    {'arange': <staticmethod object at 0x000002C60D8ABBE0>, 'as_tensor': <staticmethod object at 0x00000
32984    {'threadpoolctl-2.1.0.dist-info', 'mistune-0.8.4.dist-info', 'pluggy-0.13.1.dist-info', 'gmpy2.cp38-
32984    {'threadpoolctl-2.1.0.dist-info', 'mistune-0.8.4.dist-info', 'flask-1.1.2.dist-info', 'pluggy-0.13.1
18520    {'__name__': 'torch._C', '__doc__': None, '__package__': 'torch', '__loader__': <_frozen_importlib_e
18520    {'__len__': <slot wrapper '__len__' of 'torch._C._TensorBase' objects>, '__getitem__': <slot wrapper
18520    {'__name__': 'numpy', '__doc__': '\nNumPy\n=====\n\nProvides\n  1. An array object of arbitrary homo
18520    {'__name__': 'numpy.core', '__doc__': '\nContains the core of NumPy: ndarray, ufuncs, dtypes, etc.\n

types:
11004    <class 'function'>
6835     <class 'dict'>
6799     <class 'tuple'>
4000     <class 'method_descriptor'>
3813     <class 'builtin_function_or_method'>
3189     <class 'getset_descriptor'>
2996     <class 'weakref'>
2943     <class 'wrapper_descriptor'>
1414     <class 'type'>
1263     <class 'staticmethod'>

------------------------------------------------------------

refs:
2338    <class 'dict'> {140726060042256: <weakref at 0x000002C60B8E2DB0; to 'type' at 0x00007FFD56D1C810 (type)>, 140726060
1075    <class 'list'> [(<function _ntuple.<locals>.parse at 0x000002C6129438B0>, 'aten::_pair'), (<function _ntuple.<local
1053    <class 'dict'> (broken __repr__, type=<class 'dict'>)
885     <class 'list'> ['typename', 'is_tensor', 'is_storage', 'set_default_tensor_type', 'set_rng_state', 'get_rng_state',
731     <class 'dict'> {'arange': <staticmethod object at 0x000002C60D8ABBE0>, 'as_tensor': <staticmethod object at 0x00000
677     <class 'list'> ['\n', 'r"""\n', 'The torch package contains data structures for multi-dimensional\n', 'tensors and
626     <class 'dict'> {'sys': <module 'sys' (built-in)>, 'builtins': <module 'builtins' (built-in)>, '_frozen_importlib':
626     <class 'dict'> {'__new__': <built-in method __new__ of torch.tensortype object at 0x000002C610A63E40>, '__doc__': N
626     <class 'dict'> {'__new__': <built-in method __new__ of torch.tensortype object at 0x000002C610A64040>, '__doc__': N
626     <class 'dict'> {'__new__': <built-in method __new__ of torch.tensortype object at 0x000002C610A64240>, '__doc__': N

bytes:
36960    {140726060042256: <weakref at 0x000002C60B8E2DB0; to 'type' at 0x00007FFD56D1C810 (type)>, 140726060
36960    {'sys': <module 'sys' (built-in)>, 'builtins': <module 'builtins' (built-in)>, '_frozen_importlib':
36960    (broken __repr__, type=<class 'dict'>)
36960    {'arange': <staticmethod object at 0x000002C60D8ABBE0>, 'as_tensor': <staticmethod object at 0x00000
32984    {'threadpoolctl-2.1.0.dist-info', 'mistune-0.8.4.dist-info', 'pluggy-0.13.1.dist-info', 'gmpy2.cp38-
32984    {'threadpoolctl-2.1.0.dist-info', 'mistune-0.8.4.dist-info', 'flask-1.1.2.dist-info', 'pluggy-0.13.1
18520    {'__name__': 'torch._C', '__doc__': None, '__package__': 'torch', '__loader__': <_frozen_importlib_e
18520    {'__len__': <slot wrapper '__len__' of 'torch._C._TensorBase' objects>, '__getitem__': <slot wrapper
18520    {'__name__': 'numpy', '__doc__': '\nNumPy\n=====\n\nProvides\n  1. An array object of arbitrary homo
18520    {'__name__': 'numpy.core', '__doc__': '\nContains the core of NumPy: ndarray, ufuncs, dtypes, etc.\n

types:
11004    <class 'function'>
6831     <class 'dict'>
6798     <class 'tuple'>
4000     <class 'method_descriptor'>
3813     <class 'builtin_function_or_method'>
3189     <class 'getset_descriptor'>
2996     <class 'weakref'>
2943     <class 'wrapper_descriptor'>
1414     <class 'type'>
1263     <class 'staticmethod'>
```
As you can see, no memory leak can be spotted.

**Comment by VitalyFedyunin** [comment]
How do you measure memory usage?

**Comment by 836304831** [comment]
I used  memory_profiler to monitor run() function memeory used. becuse of I juse repeat run 3000 times, the gloabal object use memory should be stable，so using memory_profile tool to monitor run() function used memory looks fine. run each loop, I aslo manual execute gc.collection(). So, Theoretically， the sum momory is stable, but the result not meet with expectation。
The sum momory graph as follow：
![image](https://user-images.githubusercontent.com/8056281/114514661-1c4bd700-9c6e-11eb-87c1-19464aaffa05.png)
